### PR TITLE
feat: Cart API classes extendibility and inline documentation

### LINKED ISSUE #12277: Cart API classes extendibility and inline documentation
To have the Cart API extendible, the classes should be updated, so existing private methods, fields and injected dependencies are made at least protected so it is possible to access them in derived classes.

### In scope
For every class (*.ts except for *.spec.ts and *.module.ts) in
- core/cart
- storefrontlib/cms-components/cart

1a. Every function's access modifier is protected or higher
1b. Every field's access modifier is protected or higher, or getter & setter functions are provided
​
2a. Every dependency class (classes injected in constructor) are public
2b. Every class is exposed through public API (index.ts) all the way to the top of the package (cart/index.ts)

3. If the previously private method or field purpose is not understandable, provide inline documentation (code comments) or code refactors in order to have the public Cart API properly described.

### Out of scope
Refactoring of ngrx logic (effects, actions, selectors, reducers). The ngrx code should not become a part of public API.

### Epic
https://github.com/SAP/spartacus/issues/9022

### PR DESCRIPTION
**Pull request by marcinkapica**
Existing private methods, fields and injected dependencies in Cart API classes (classes in core/cart and storefrontlib/cms-components/cart, excluding ngrx store-related code) were made protected to make the Cart API more extendible. Modified methods were refactored, updated with method descriptions and inline comments where necessary.

Closes #12277

### CODE REVIEW

File: projects/core/src/cart/facade/active-cart.service.ts [hunk 1]
```diff
@@ -220,15 +221,17 @@ export class ActiveCartService implements OnDestroy {
...
+  protected loadOrMerge(
     cartId: string,
     userId: string,
     previousUserId: string
   ): void {
-    // for login user, whenever there's an existing cart, we will load the user
-    // current cart and merge it into the existing cart
-    // cartId will be defined (not '', null, undefined)
     if (cartId === OCC_CART_ID_CURRENT) {
+      // Load cart if it is current
```
Comment (marlass) [inline projects/core/src/cart/facade/active-cart.service.ts]: I don't know if this comment brings any value ;)

File: projects/core/src/cart/facade/active-cart.service.ts [hunk 2]
```diff
@@ -237,6 +240,7 @@ export class ActiveCartService implements OnDestroy {
         },
       });
     } else if (this.isGuestCart()) {
+      // If the cart is guest cart, merge it with user card
```
Comment (marlass) [inline projects/core/src/cart/facade/active-cart.service.ts]: Same with this. It's quite obvious from code

File: projects/core/src/cart/facade/cart-voucher.service.ts [hunk 3]
```diff
@@ -89,7 +89,13 @@ export class CartVoucherService {
...
   }
 
-  private combineUserAndCartId(cartId: string): Observable<[string, string]> {
+  /**
+   * Returns observable with array containing user id and cart id
+   * @param cartId optional parameter with cart id - if not provided, id of active cart will be used
+   */
+  protected combineUserAndCartId(
+    cartId?: string
+  ): Observable<[string, string]> {
```
Comment (marlass) [inline projects/core/src/cart/facade/cart-voucher.service.ts]: This method is too ugly and not really useful to even be in public API. Probably even candidate for removal ;)

File: projects/core/src/cart/facade/active-cart.service.ts (hunk 1 above)
Comment (marcinkapica) [reply projects/core/src/cart/facade/active-cart.service.ts]: Comment removed

File: projects/core/src/cart/facade/active-cart.service.ts (hunk 2 above)
Comment (marcinkapica) [reply projects/core/src/cart/facade/active-cart.service.ts]: Comment removed

File: projects/core/src/cart/facade/cart-voucher.service.ts (hunk 3 above)
Comment (marcinkapica) [reply projects/core/src/cart/facade/cart-voucher.service.ts]: Ok, changed back to private as it was

**Review by marlass — APPROVED** [review]

**Pull request merged**